### PR TITLE
Fix 7zip building on mingw on Windows

### LINKED ISSUE #394: [question][package] How to compile 7zip in GCC 9.2.0 on Windows?
I'm trying to compile 7zip with the new GCC 9.2.0 from STL.

I created a very simple conanfile.txt:
```
[requires]
7zip/19.00

[generators]
cmake
```

After executing:
`conan install .. -pr mingw32 -b=missing`

I get the following error:
```
C:\Users\User\Documents\Sources\testing\conan>conan install . -pr mingw32 -b=missing
Configuration:
[settings]
arch=x86
arch_build=x86
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=9
os=Windows
os_build=Windows
[options]
[build_requires]
openssl*: msys2/20161025
[env]
CC=C:\MinGW\bin\gcc.exe
CXX=C:\MinGW\bin\g++.exe
7zip/19.00: WARN: Package is corrupted, removing folder: C:\Users\User\.conan\data\7zip\19.00\_\_\package\b2fcbc60b3bee75023a4f7934cadb723eec550bd
conanfile.txt: Installing package
Requirements
    7zip/19.00 from 'conan-center' - Cache
Packages
    7zip/19.00:b2fcbc60b3bee75023a4f7934cadb723eec550bd - Build

7zip/19.00: WARN: Build folder is dirty, removing it: C:\Users\User\.conan\data\7zip\19.00\_\_\build\b2fcbc60b3bee75023a4f7934cadb723eec550bd
7zip/19.00: Copying sources to build folder
7zip/19.00: Building your package in C:\Users\User\.conan\data\7zip\19.00\_\_\build\b2fcbc60b3bee75023a4f7934cadb723eec550bd
7zip/19.00: Generator txt created conanbuildinfo.txt
7zip/19.00: Calling build()
g++ -O2 -c -D_7ZIP_ST ../../../Common/C_FileIO.cpp
g++ -O2 -c -D_7ZIP_ST LzmaAlone.cpp
g++ -O2 -c -D_7ZIP_ST ../../UI/Common/Bench.cpp
g++ -O2 -c -D_7ZIP_ST ../../UI/Console/BenchCon.cpp
g++ -O2 -c -D_7ZIP_ST ../../UI/Console/ConsoleClose.cpp
g++ -O2 -c -D_7ZIP_ST ../../Compress/LzmaDecoder.cpp
g++ -O2 -c -D_7ZIP_ST ../../Compress/LzmaEncoder.cpp
g++ -O2 -c -D_7ZIP_ST ../../Compress/LzmaRegister.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/CreateCoder.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/CWrappers.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/FileStreams.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/FilterCoder.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/MethodProps.cpp
g++ -O2 -c -D_7ZIP_ST ../../Common/StreamUtils.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/CommandLineParser.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/CRC.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/CrcReg.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/IntToString.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/MyString.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/MyVector.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/MyWindows.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/StringConvert.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Common/StringToInt.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Windows/PropVariant.cpp
g++ -O2 -c -D_7ZIP_ST ../../../Windows/System.cpp
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/7zCrc.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/7zCrcOpt.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/Alloc.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/Bra86.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/CpuArch.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/LzFind.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/LzmaDec.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/LzmaEnc.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/Lzma86Dec.c
gcc -O2 -Wall -Werror -c -D_7ZIP_ST ../../../../C/Lzma86Enc.c
g++ -O2 -o lzma   C_FileIO.o LzmaAlone.o Bench.o BenchCon.o ConsoleClose.o LzmaDecoder.o LzmaEncoder.o LzmaRegister.o CreateCoder.o CWrappers.o FileStreams.o FilterCoder.o MethodProps.o StreamUtils.o CommandLineParser.o CRC.o CrcReg.o IntToString.o MyString.o MyVector.o MyWindows.o StringConvert.o StringToInt.o PropVariant.o System.o 7zCrc.o 7zCrcOpt.o Alloc.o Bra86.o CpuArch.o LzFind.o LzmaDec.o LzmaEnc.o Lzma86Dec.o Lzma86Enc.o
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: LzmaAlone.o:LzmaAlone.cpp:(.text+0x92a): undefined reference to `NWindows::NFile::NIO::CFileBase::GetLength(unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: LzmaAlone.o:LzmaAlone.cpp:(.text+0xc8e): undefined reference to `NWindows::NFile::NIO::CInFile::Open(wchar_t const*)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: LzmaAlone.o:LzmaAlone.cpp:(.text+0xd00): undefined reference to `NWindows::NFile::NIO::COutFile::Create(wchar_t const*, bool)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: LzmaAlone.o:LzmaAlone.cpp:(.rdata$.refptr.IID_IUnknown[.refptr.IID_IUnknown]+0x0): undefined reference to `IID_IUnknown'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x44a9): undefined reference to `NWindows::NFile::NIO::CInFile::Open(wchar_t const*)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x44c4): undefined reference to `NWindows::NFile::NIO::CFileBase::GetLength(unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x453d): undefined reference to `NWindows::NFile::NIO::CInFile::Read(void*, unsigned int, unsigned int&)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x458c): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x4cfa): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: Bench.o:Bench.cpp:(.text+0x6995): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x220): undefined reference to `NWindows::NFile::NIO::CFileBase::Seek(long long, unsigned long, unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x22f): undefined reference to `NWindows::NFile::NIO::COutFile::SetLength(unsigned long long)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x245): undefined reference to `NWindows::NFile::NIO::CFileBase::Seek(unsigned long long, unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x286): undefined reference to `NWindows::NFile::NIO::CInFile::ReadPart(void*, unsigned int, unsigned int&)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x3c9): undefined reference to `NWindows::NFile::NIO::CFileBase::GetLength(unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x438): undefined reference to `NWindows::NFile::NIO::CFileBase::Seek(long long, unsigned long, unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x4a8): undefined reference to `NWindows::NFile::NIO::CFileBase::Seek(long long, unsigned long, unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x516): undefined reference to `NWindows::NFile::NIO::COutFile::Write(void const*, unsigned int, unsigned int&)'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x669): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x6b9): undefined reference to `NWindows::NFile::NIO::CFileBase::GetLength(unsigned long long&) const'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text+0x47): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text$_ZN14COutFileStreamD1Ev[_ZN14COutFileStreamD1Ev]+0xf): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text$_ZN14COutFileStreamD0Ev[_ZN14COutFileStreamD0Ev]+0x18): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: FileStreams.o:FileStreams.cpp:(.text$_ZN14COutFileStream7ReleaseEv[_ZN14COutFileStream7ReleaseEv]+0x3e): undefined reference to `NWindows::NFile::NIO::CFileBase::Close()'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: MyString.o:MyString.cpp:(.text+0x2026): undefined reference to `__imp_SysStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x1c): undefined reference to `__imp_SysAllocStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x7c): undefined reference to `__imp_SysAllocStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x162): undefined reference to `__imp_VariantClear'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x1d5): undefined reference to `__imp_SysAllocString'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x258): undefined reference to `__imp_SysAllocStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x2c8): undefined reference to `__imp_SysAllocStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x3e8): undefined reference to `__imp_SysAllocStringLen'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x600): undefined reference to `__imp_VariantClear'
c:/mingw/bin/../lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: PropVariant.o:PropVariant.cpp:(.text+0x630): undefined reference to `__imp_VariantCopy'
collect2.exe: error: ld returned 1 exit status
make: *** [makefile.gcc:78: lzma] Error 1
7zip/19.00:
7zip/19.00: ERROR: Package 'b2fcbc60b3bee75023a4f7934cadb723eec550bd' build failed
7zip/19.00: WARN: Build folder C:\Users\User\.conan\data\7zip\19.00\_\_\build\b2fcbc60b3bee75023a4f7934cadb723eec550bd
ERROR: 7zip/19.00: Error in build() method, line 62
        self._build_autotools()
while calling '_build_autotools', line 56
        self.run("make -f makefile.gcc all")
        ConanException: Error 2 while executing make -f makefile.gcc all
```
It looks like the linker is not recognizing the object files properly and it also links to win32api available in MinGW.

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **7zip/19.00**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

This should fix #394

- This will force the msvc runtime by patching the sources (a lot of warnings were emitted during build)
- This will set the `IS_MINGW` environment variable if mingw is detected

### CODE REVIEW

**Comment by madebr** [comment]
This will fail because it needs #396

File: recipes/7zip/19.00/conanfile.py [hunk 1]
```diff
@@ -36,26 +39,36 @@ def _uncompress_7z(self, filename):
...
     def _build_autotools(self):
         # TODO: Enable non-Windows methods in configure
-        env_build = AutoToolsBuildEnvironment(self)
-        with tools.environment_append(env_build.vars):
-            with tools.chdir("CPP/7zip/Bundles/LzmaCon"):
-                self.run("make -f makefile.gcc all")
+        autotools = AutoToolsBuildEnvironment(self)
+        extra_env = {}
+        if self.settings.os_build == "Windows" and self.settings.compiler == "gcc":
+            extra_env["IS_MINGW"] = "1"
```
Comment (jgsogo) [inline recipes/7zip/19.00/conanfile.py]: I'd add here a link to the issue https://github.com/conan-io/conan-center-index/issues/394, in the future we should fix it in Conan (https://github.com/conan-io/conan/issues/1706) and come back to the recipe to remove this patch.
```suggestion
            extra_env["IS_MINGW"] = "1"  # FIXME: https://github.com/conan-io/conan/issues/1706
```

**Comment by danimtb** [comment]
Blocked by https://github.com/conan-io/conan-center-index/issues/571

**Comment by danimtb** [comment]
relaunched

**Pull request merged**